### PR TITLE
Cypress IO end to end tests

### PR DESCRIPTION
**Pull request by fjgarlin**
Fixes https://github.com/Lullabot/drupal9ci/issues/67

Work in progress. Added base files and jobs for all platforms. 
**Nothing tested yet**, just wanted to upload progress with the implementation that I had in mind.

- [x] Gitlab: https://gitlab.com/fjgarlin/d9-lagoon/-/pipelines/286151085
- [x] Github actions: https://github.com/fjgarlin/d9-lagoon/actions/runs/753078876
- [x] CircleCI: https://github.com/fjgarlin/d9-lagoon/pull/8 (see last √)
- [x] TravisCI: https://travis-ci.com/github/fjgarlin/d9-lagoon/builds/223392448

### CODE REVIEW

File: dist/circleci/.circleci/RoboFile.php [hunk 1]
```diff
@@ -139,6 +156,26 @@ protected function runBehatTests()
...
+     */
+    protected function runCypressTests()
+    {
+        $force = TRUE;
+        $tasks = [];
+        $tasks[] = $this->taskExec('service apache2 start');
+        $tasks[] = $this->taskFilesystemStack()
+            ->copy('.cypress/cypress.json', 'cypress.json', $force)
+            ->copy('.cypress/package.json', 'package.json', $force);
+        $tasks[] = $this->taskExec('sleep 30s');
```
Comment (juampynr) [inline dist/circleci/.circleci/RoboFile.php:173]: Why the sleep? Just curious.

File: dist/circleci/.circleci/config.yml [hunk 2]
```diff
@@ -51,6 +48,26 @@ behat_tests: &behat_tests
     - store_artifacts:
         path: /opt/drupal/artifacts
 
+## Job to run the update path and Cypress tests.
+cypress_tests: &cypress_tests
+  <<: *defaults
+  docker:
+    - image: juampynr/drupal9ci:latest
+    - image: selenium/standalone-chrome-debug:3.7.1-beryllium
```
Comment (juampynr) [inline dist/circleci/.circleci/config.yml:56]: My understanding is that cypress uses its own browser so selenium is not needed, nor chromedriver, nor chrome headless browser.

**Comment by juampynr** [comment]
Looks good so far 👍🏽

File: dist/circleci/.circleci/config.yml (hunk 2 above)
Comment (fjgarlin) [reply dist/circleci/.circleci/config.yml:56]: Well spotted, you're right, this is not needed. Probably came from a copy/paste.

File: dist/circleci/.circleci/RoboFile.php (hunk 1 above)
Comment (fjgarlin) [reply dist/circleci/.circleci/RoboFile.php:173]: With behat, for some of the platforms, as starting the server was run in parallel as the behat task, that sleep made sure that the server was up and running. 

With cypress, we might or might not have the same situation. Probably this is just copy/paste as I based it on the behat set up, but if might not be needed on all platforms.

**Comment by fjgarlin** [comment]
Gitlab Cypress is on! 💪  https://gitlab.com/fjgarlin/d9-lagoon/-/pipelines/286151085

**Comment by juampynr** [comment]
@fjgarlin let me know when this is ready for another review.

**Comment by fjgarlin** [comment]
Github actions in _action_: https://github.com/fjgarlin/d9-lagoon/actions/runs/753078876

**Comment by fjgarlin** [comment]
@juampynr - I think this is now ready for review. TravisCI took me a bit longer mostly because I wanted to simplify the `docker-compose` architecture to be in line with all the others. It now uses your images to build the docker containers and test behat and cypress.

All four platforms had successful CI passes and are linked at the top of this PR for reference.

File: dist/common/.cypress/README.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+dependencies. If you want to test locally, you can copy this file to the
+root of the repo (if you don't have one already) and just run locally
+`npm install cypress --save-dev`.
+
+`cypress.json` is set up to work with the CI integration. If you want to
+test locally, copy this file to the root of the repo and tweak any of the
+values if needed (ie: `baseUrl`).
+
+Once everything is set up locally, you can run the tests like this:
+`$(npm bin)/cypress open` or `$(npm bin)/cypress run`.
```
Comment (juampynr) [inline dist/common/.cypress/README.md:14]: This looks great.

File: dist/github-actions/.github/workflows/ci.yml [hunk 4]
```diff
@@ -105,3 +105,38 @@ jobs:
         run: |
           vendor/bin/robo job:serve-drupal &
           vendor/bin/robo job:behat-tests
+
+  cypress:
+    runs-on: ubuntu-latest
+    container:
+      image: juampynr/drupal8ci:latest
```
Comment (juampynr) [inline dist/github-actions/.github/workflows/ci.yml:112]: Does this need the drupal8 image or could it use the drupal9 one?

**Comment by juampynr** [comment]
Posted a few comments but this is already a huge step forward so I am merging it @fjgarlin.

THANK YOU!

File: dist/github-actions/.github/workflows/ci.yml (hunk 4 above)
Comment (fjgarlin) [reply dist/github-actions/.github/workflows/ci.yml:112]: It depends on the PHP version requirements. For projects using PHP7.3 or PHP7.4, the `drupal8ci` CI image should be used. For projects using PHP8.0 then the `drupal9ci` is good.

**Comment by fjgarlin** [comment]
You're welcome, it's been fun to bring Cypress into the combo. This whole project is really cool!

**Pull request merged**